From a85da35ad8fb2f446fceace4c081edf8d34bb572 Mon Sep 17 00:00:00 2001
From: Zhantao Tang <zhantao.tang@windriver.com>
Date: Mon, 1 Mar 2021 15:51:54 +0800
Subject: [PATCH] pkcs: example: modify the example codes to workable

In the example source codes, the keyid[] array related
to the new generated RSA key pair should set to:
        keyid[0] = 1; /* slot ID */
        keyid[1] = 5; /* group ID */
        keyid[2] = 1; /* catalog ID */

then the example app should work.
And there add some codes for convenient to try different
keyid[] parameters.

Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>
---
 examples/pkcs-keyop.c | 22 +++++++++++++++++-----
 1 file changed, 17 insertions(+), 5 deletions(-)

diff --git a/examples/pkcs-keyop.c b/examples/pkcs-keyop.c
index 5b73e09..608e084 100644
--- a/examples/pkcs-keyop.c
+++ b/examples/pkcs-keyop.c
@@ -10,6 +10,7 @@
  */
 
 #include <stdio.h>
+#include <stdlib.h>
 #include <libp11.h>
 
 int main(int argc, char *argv[])
@@ -23,12 +24,23 @@ int main(int argc, char *argv[])
 	unsigned char keyid[3];
 	PKCS11_KEY *keyp;
 	unsigned int keycount;
+	int catID = 1;
+	int groupID = 5;
+	int slotID = 1;
 
-	if (argc != 2) {
-		printf("usage: ./pkcs-keyop <path>/pkcs11-nxp.so\n");
+	if (argc < 2) {
+		printf("usage: ./pkcs-keyop <path>/libpkcs11-hse.so [option catID] [option groupID] [option slotID] \n");
+		printf("\texample: ./pkcs-keyop <path>/libpkcs11-hse.so 1 5 1 \n");
 		return -EINVAL;
 	}
 
+	if (argc >=3)
+		catID = atoi(argv[2]);
+	if (argc >=4)
+		groupID = atoi(argv[3]);
+	if (argc >=5)
+		slotID = atoi(argv[4]);
+
 	/* create a context for loading pkcs11 module */
 	pkcs_ctx = PKCS11_CTX_new();
 
@@ -89,9 +101,9 @@ int main(int argc, char *argv[])
 	 *     - keyid[0] - slot ID
 	 * HSE key IDs can only be composed of 3 bytes
 	 */
-	keyid[0] = 0x01;
-	keyid[1] = 0x06;
-	keyid[2] = 0x01;
+	keyid[0] = slotID;
+	keyid[1] = groupID;
+	keyid[2] = catID;
 
 	/* 
 	 * store the the key pair
-- 
2.25.1

